handshake_ix.go

```diff
@@ -68,28 +67,51 @@ func ixHandshakeStage0(f *Interface, vpnIp uint32, hostinfo *HostInfo) {
 }
 
 func ixHandshakeStage1(f *Interface, addr *udpAddr, packet []byte, h *Header) {
-	ci := f.newConnectionState(f.l, false, noise.HandshakeIX, []byte{}, 0)
-	// Mark packet 1 as seen so it doesn't show up as missed
-	ci.window.Update(f.l, 1)
+	var (
+		err error
+		ci  *ConnectionState
+		msg []byte
+	)
 
-	msg, _, _, err := ci.H.ReadMessage(nil, packet[HeaderLen:])
-	if err != nil {
-		f.l.WithError(err).WithField("udpAddr", addr).
-			WithField("handshake", m{"stage": 1, "style": "ix_psk0"}).Error("Failed to call noise.ReadMessage")
-		return
+	hs := &NebulaHandshake{}
+
+	// Handle multiple possible psk options, ensure the protobuf comes out clean too
+	for _, p := range f.psk.Cache {
+		//TODO: benchmark generation time of makePsk
+		ci, err = f.newConnectionState(f.l, false, p)
+		if err != nil {
+			f.l.WithError(err).WithField("udpAddr", addr).Error("Failed to get a new connection state")
+			continue
+		}
+
+		msg, _, _, err = ci.H.ReadMessage(nil, packet[HeaderLen:])
+		if err != nil {
+			// Calls to ReadMessage with an incorrect psk should fail, try the next one if we have one
+			continue
+		}
+
+		// Sometimes ReadMessage returns fine with a nil psk even if the handshake is using a psk, ensure our protobuf
+		// comes out clean as well
```

Comment:
I'm curious to understand more about what conditions would cause this to occur - if you know, expanding on the code comment might help future readers.

Reply:
After spending a few hours reading about Noise, I think this code comment applies to when PSK is set to `none` or otherwise set to an empty value. Since `k` won't be set to anything yet, we'll simply return out whatever data we get here: https://github.com/flynn/noise/blob/master/state.go#L134-L137

When we have PSK turned on, our `k` will be seeded, and we won't hit the conditional. If we have PSK turned off, and the other side has PSK turned on, I believe we would hit this condition. Or if we have PSK turned off and the other side is just sending garbage.